File: mavsim_matlab/chap12/rrt_dubins.m
% rrt dubins path planner for mavsim_matlab
%
% mavsim_matlab 
%     - Beard & McLain, PUP, 2012
%     - Last updated:  
%         4/10/2019 - RWB
classdef rrt_dubins < handle
   %--------------------------------
    properties
        segment_length
        tree
        dubins_path
    end
    %--------------------------------
    methods
        %------constructor-----------
        function self = rrt_dubins
            % standard length of path segments
            self.segment_length = 600;
            % waypoints definition
            addpath('../message_types'); 
            self.tree = msg_waypoints();
            self.tree.type = 'dubins';
            addpath('../chap11')
            self.dubins_path = dubins_parameters();
        end
        %------methods-----------
        function waypoints = update(self, start_pose, end_pose, Va, map, radius)
            % add the start node to the tree
            self.tree.add(start_pose(1:3), Va, start_pose(4), 0, 0, 0);
            % add(ned_position, airspeed, course, cost, flag_connect_to_goal)

            % check to see if start_node connects directly to end_node
            if ( (self.distance(start_pose, end_pose)<self.segment_length )...
                    && (self.collision(start_pose, end_pose, map, radius)==0)...
                    && (self.distance(start_pose, end_pose)>=2*radius) )
                self.tree.add(end_pos, Va, inf, self.distance(start_pos, end_pos), 1, 1);
            else
                numPaths = 0;
                while numPaths<3
                    flag = self.extend_tree(end_pose, Va, map, radius);
                    numPaths = numPaths + flag;
                end
            end

            % find path with minimum cost to end_node
            waypoints_not_smooth = self.find_minimum_path(end_pose);
            waypoints = self.smooth_path(waypoints_not_smooth, map, radius);
            self.plot_map(map, waypoints_not_smooth, waypoints, radius);
        end
        %----------------------------------------
        function d=distance(self, start_pose, end_pose)
        end
        %----------------------------------------
        function collision_flag = collision(self, pose_s, pose_e, map, radius)
        end
        %----------------------------------------
        function points = points_along_path(self, Del)
            %   Find points along Dubins path separted by Del (to be used in
            %   collision detection)
            points = [];
            % points along start circle
            th1 = mod(atan2(self.dubins_path.p_s(2)-self.dubins_path.center_s(2),...
                            self.dubins_path.p_s(1)-self.dubins_path.center_s(1)), 2*pi);
            th2 = mod(atan2(self.dubins_path.r1(2)-self.dubins_path.center_s(2),...
                            self.dubins_path.r1(1)-self.dubins_path.center_s(1)), 2*pi);
            if self.dubins_path.dir_s>0
                if th1>=th2
                    th = [th1:Del:2*pi,0:Del:th2];
                else
                    th = [th1:Del:th2];
                end
            else
                if th1<=th2
                    th = [th1:-Del:0,2*pi:-Del:th2];
                else
                    th = [th1:-Del:th2];
                end
            end
            for i=1:length(th)
                new_point = [...
                    self.dubins_path.center_s(1)+self.dubins_path.radius*cos(th(i));...
                    self.dubins_path.center_s(2)+self.dubins_path.radius*sin(th(i));...
                    self.dubins_path.center_s(3);...
                    ];
                points = [points, new_point];
            end
  
            % points along straight line 
            sig = 0;
            while sig<=1
                new_point = [...
                    (1-sig)*self.dubins_path.r1(1) + sig*self.dubins_path.r2(1);...
                    (1-sig)*self.dubins_path.r1(2) + sig*self.dubins_path.r2(2);...
                    (1-sig)*self.dubins_path.r1(3) + sig*self.dubins_path.r2(3);...
                    ];
                points = [points, new_point];
                sig = sig + Del;
            end
    
            % points along end circle
            th2 = mod(atan2(self.dubins_path.p_e(2)-self.dubins_path.center_e(2),...
                            self.dubins_path.p_e(1)-self.dubins_path.center_e(1)), 2*pi);
            th1 = mod(atan2(self.dubins_path.r2(2)-self.dubins_path.center_e(2),...
                            self.dubins_path.r2(1)-self.dubins_path.center_e(1)), 2*pi);
            if self.dubins_path.dir_e>0
                if th1>=th2
                    th = [th1:Del:2*pi,0:Del:th2];
                else
                    th = [th1:Del:th2];
                end
            else
                if th1<=th2
                    th = [th1:-Del:0,2*pi:-Del:th2];
                else
                    th = [th1:-Del:th2];
                end
            end
            for i=1:length(th)
                new_point = [...
                    self.dubins_path.center_e(1)+self.dubins_path.radius*cos(th(i));...
                    self.dubins_path.center_e(2)+self.dubins_path.radius*sin(th(i));...
                    self.dubins_path.center_e(3);...
                    ];
                points = [points, new_point];
            end
        end
        %----------------------------------------
        function h_agl = height_above_ground(self, map, point)
            %   finds altitude above ground level
        end
        %----------------------------------------
        function flag = extend_tree(self, end_pose, Va, map, radius)
        end
        %----------------------------------------
        function pose=random_pose(self, map, pd)        
        end

        %----------------------------------------
        function waypoints = find_minimum_path(self, end_node)
        end
        %----------------------------------------
        function smooth_waypoints = smooth_path(self, waypoints, map, radius)
        end
        %----------------------------------------
        function self = plot_map(self, map, waypoints, smoothed_waypoints, radius)
            addpath('../chap12'); world_view = world_viewer();
            S = 2000;
            figure(3); clf;
            world_view.drawMap(map);
            hold on
            world_view.drawNonSmoothedWaypoints(waypoints, radius);
            world_view.drawWaypoints(smoothed_waypoints, radius);
            world_view.drawTree(self.tree, radius);
            
            title('UAV')
            xlabel('East')
            ylabel('North')
            zlabel('-Down')
            axis([-S/5,S+S/5,-S/5,S+S/5,0,3*map.building_max_height]);
            view(-40,70)  % set the view angle for figure
            axis square;
            grid on
        end
    end
end

File: legacy_mavsim_python/chap12/rrt_dubins.py
# rrt dubins path planner for mavsim_python
#
# mavsim_python
#     - Beard & McLain, PUP, 2012
#     - Last updated:
#         4/16/2019 - RWB
import numpy as np
from message_types.msg_waypoints import MsgWaypoints
from chap11.draw_waypoints import DrawWaypoints
from chap12.draw_map import DrawMap
from chap11.dubins_parameters import DubinsParameters
import pyqtgraph as pg
import pyqtgraph.opengl as gl


class RRTDubins:
    def __init__(self):
        self.segment_length = 300  # standard length of path segments
        self.plot_window = []
        self.plot_app = []
        self.dubins_path = DubinsParameters()

    def update(self, start_pose, end_pose, Va, world_map, radius):
        #generate tree
        tree = MsgWaypoints()
        tree.type = 'dubins'
        # add the start pose to the tree
        
        # check to see if start_pose connects directly to end_pose
        
        # find path with minimum cost to end_node
        waypoints_not_smooth = #find_minimum_path()
        waypoints = #smooth_path()

        return waypoints

    def extend_tree(self, tree, end_pose, Va, world_map, radius):
        # extend tree by randomly selecting pose and extending tree toward that pose
        flag = 
        return flag

    def points_along_path(self, Del):
        # returns a list of points along the dubins path
        initialize_points = True
        # points along start circle
        theta_list = []
        if initialize_points:
            points = 
            initialize_points = False
        for angle in theta_list:
            points = 

        # points along straight line
        sig = 0
        while sig <= 1:
            points =
            sig = 

        # points along end circle
    
        theta_list = []
        for angle in theta_list:
            points = 
        return points

    def collision(self, start_pose, end_pose, world_map, radius):
        # check to see of path from start_pose to end_pose colliding with world_map
        collision_flag = 
        return collision_flag

    def plot_map(self, world_map, tree, waypoints, smoothed_waypoints, radius):
        scale = 4000
        # initialize Qt gui application and window
        self.plot_app = pg.QtGui.QApplication([])  # initialize QT
        self.plot_window = gl.GLViewWidget()  # initialize the view object
        self.plot_window.setWindowTitle('World Viewer')
        self.plot_window.setGeometry(0, 0, 1500, 1500)  # args: upper_left_x, upper_right_y, width, height
        grid = gl.GLGridItem() # make a grid to represent the ground
        grid.scale(scale/20, scale/20, scale/20) # set the size of the grid (distance between each line)
        self.plot_window.addItem(grid) # add grid to viewer
        self.plot_window.setCameraPosition(distance=scale, elevation=50, azimuth=-90)
        self.plot_window.setBackgroundColor('k')  # set background color to black
        self.plot_window.show()  # display configured window
        self.plot_window.raise_() # bring window to the front

        blue = np.array([[30, 144, 255, 255]])/255.
        red = np.array([[204, 0, 0]])/255.
        green = np.array([[0, 153, 51]])/255.
        DrawMap(world_map, self.plot_window)
        DrawWaypoints(waypoints, radius, blue, self.plot_window)
        DrawWaypoints(smoothed_waypoints, radius, red, self.plot_window)
        self.draw_tree(radius, green)
        # draw things to the screen
        self.plot_app.processEvents()

    def draw_tree(self, radius, color, window):
        R = np.array([[0, 1, 0], [1, 0, 0], [0, 0, -1]])
        Del = 0.05
        for i in range(1, tree.num_waypoints):
            parent = int(tree.parent.item(i))
            self.dubins_path.update(column(tree.ned, parent), tree.course[parent],
                                    column(tree.ned, i), tree.course[i], radius)
            points = self.points_along_path(Del)
            points = points @ R.T
            tree_color = np.tile(color, (points.shape[0], 1))
            tree_plot_object = gl.GLLinePlotItem(pos=points,
                                                color=tree_color,
                                                width=2,
                                                antialias=True,
                                                mode='line_strip')
            self.plot_window.addItem(tree_plot_object)


def find_minimum_path(tree, end_pose):
    # find the lowest cost path to the end node
    # find nodes that connect to end_node
    connecting_nodes = 

    # find minimum cost last node
    idx = 

    # construct lowest cost path order
    path =   # last node that connects to end node

    # construct waypoint path
    waypoints = MsgWaypoints()

    return waypoints


def smooth_path(waypoints, world_map, radius):
    # smooth the waypoint path
    smooth = [0]  # add the first waypoint

    # construct smooth waypoint path
    smooth_waypoints = MsgWaypoints()
    
    return smooth_waypoints


def distance(start_pose, end_pose):
    # compute distance between start and end pose
    d = 
    return d


def height_above_ground(world_map, point):
    # find the altitude of point above ground level
    point_height = 
    if (d_n<world_map.building_width) and (d_e<world_map.building_width):
        map_height = 
    else:
        map_height = 
    h_agl = 
    return h_agl


def random_pose(world_map, pd):
    # generate a random pose
    pose = 
    return pose


def mod(x):
    # force x to be between 0 and 2*pi
    while x < 0:
        x += 2*np.pi
    while x > 2*np.pi:
        x -= 2*np.pi
    return x


def column(A, i):
    # extracts the ith column of A and return column vector
    tmp = A[:, i]
    col = tmp.reshape(A.shape[0], 1)
    return col